// @nutui/nutui/dist/packages/_es/Noticebar.js
import { h, ref, reactive, computed, watch, onMounted, onActivated, onDeactivated, onUnmounted, toRefs, resolveComponent, openBlock, createElementBlock, normalizeClass, withDirectives, normalizeStyle, createElementVNode, renderSlot, createBlock, createCommentVNode, createTextVNode, toDisplayString, withModifiers, vShow, Fragment, renderList } from "vue";
import { Notice, CircleClose } from "@nutui/icons-vue";
import { c as createComponent } from "./component-81a4c1d0.js";
import { p as pxCheck } from "./pxCheck-c6b9f6b7.js";
import { _ as _export_sfc } from "./_plugin-vue_export-helper-cc2b3d55.js";
import "../locale/lang";
const { componentName, create } = createComponent("noticebar");
const _sfc_main = create({
  props: {
    direction: {
      type: String,
      default: "across"
    },
    list: {
      type: Array,
      default: () => {
        return [];
      }
    },
    standTime: {
      type: Number,
      default: 1e3
    },
    complexAm: {
      type: Boolean,
      default: false
    },
    height: {
      type: Number,
      default: 40
    },
    text: {
      type: String,
      default: ""
    },
    closeMode: {
      type: Boolean,
      default: false
    },
    wrapable: {
      type: Boolean,
      default: false
    },
    leftIcon: { type: Boolean, default: true },
    color: {
      type: String,
      default: ""
    },
    background: {
      type: String,
      default: ""
    },
    delay: {
      type: [String, Number],
      default: 1
    },
    scrollable: {
      type: Boolean,
      default: null
    },
    speed: {
      type: Number,
      default: 50
    }
  },
  components: {
    ScrollItem: function(props) {
      props.item.props.style = props.style;
      props.item.key = props.key;
      return h(props.item);
    },
    Notice,
    CircleClose
  },
  emits: ["click", "close", "across-end"],
  setup(props, { emit, slots }) {
    const wrap = ref(null);
    const content = ref(null);
    const state = reactive({
      wrapWidth: 0,
      firstRound: true,
      duration: 0,
      offsetWidth: 0,
      showNoticebar: true,
      animationClass: "",
      animate: false,
      scrollList: [],
      distance: 0,
      timer: null,
      keepAlive: false,
      isCanScroll: null,
      showNotica: true
    });
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    const isEllipsis = computed(() => {
      if (state.isCanScroll == null) {
        return props.wrapable;
      } else {
        return !state.isCanScroll && !props.wrapable;
      }
    });
    const wrapContentClass = computed(() => {
      return {
        "nut-noticebar__page-wrap-content": true,
        "nut-ellipsis": isEllipsis.value,
        [state.animationClass]: true
      };
    });
    const barStyle = computed(() => {
      let style = {};
      props.color && (style.color = props.color);
      props.background && (style.background = props.background);
      if (props.direction == "vertical") {
        style.height = `${props.height}px`;
      }
      return style;
    });
    const contentStyle = computed(() => {
      return {
        animationDelay: (state.firstRound ? props.delay : 0) + "s",
        animationDuration: state.duration + "s",
        transform: `translateX(${state.firstRound ? 0 : state.wrapWidth + "px"})`
      };
    });
    const horseLampStyle = computed(() => {
      let styles = {};
      if (props.complexAm) {
        styles = {
          transform: `translateY(${state.distance}px)`
        };
      } else {
        if (state.animate) {
          let a = ~~(props.height / props.speed / 4);
          styles = {
            transition: `all ${a == 0 ? ~~(props.height / props.speed) : a}s`,
            "margin-top": `-${props.height}px`
          };
        }
      }
      return styles;
    });
    watch(
      () => props.text,
      (value) => {
        initScrollWrap();
      }
    );
    watch(
      () => props.list,
      (value) => {
        state.scrollList = [].concat(value);
      }
    );
    const initScrollWrap = (value) => {
      if (state.showNoticebar == false) {
        return;
      }
      setTimeout(() => {
        if (!wrap.value || !content.value) {
          return;
        }
        const wrapWidth = wrap.value.getBoundingClientRect().width;
        const offsetWidth = content.value.getBoundingClientRect().width;
        state.isCanScroll = props.scrollable == null ? offsetWidth > wrapWidth : props.scrollable;
        if (state.isCanScroll) {
          state.wrapWidth = wrapWidth;
          state.offsetWidth = offsetWidth;
          state.duration = offsetWidth / props.speed;
          state.animationClass = "play";
        } else {
          state.animationClass = "";
        }
      }, 0);
    };
    const handleClick = (event) => {
      emit("click", event);
    };
    const onClickIcon = (event) => {
      if (props.closeMode) {
        state.showNoticebar = !props.closeMode;
      }
      emit("close", event);
    };
    const onAnimationEnd = (event) => {
      state.firstRound = false;
      emit("across-end", event);
      setTimeout(() => {
        state.duration = (state.offsetWidth + state.wrapWidth) / props.speed;
        state.animationClass = "play-infinite";
      }, 0);
    };
    const startRollEasy = () => {
      showhorseLamp();
      state.timer = setInterval(showhorseLamp, ~~(props.height / props.speed / 4 * 1e3) + props.standTime);
    };
    const showhorseLamp = () => {
      state.animate = true;
      setTimeout(() => {
        state.scrollList.push(state.scrollList[0]);
        state.scrollList.shift();
        state.animate = false;
      }, ~~(props.height / props.speed / 4 * 1e3));
    };
    const startRoll = () => {
      state.timer = setInterval(() => {
        let chunk = 100;
        for (let i = 0; i < chunk; i++) {
          scroll(i, i < chunk - 1 ? false : true);
        }
      }, props.standTime + 100 * props.speed);
    };
    const scroll = (n, last) => {
      setTimeout(() => {
        state.distance -= props.height / 100;
        if (last) {
          state.scrollList.push(state.scrollList[0]);
          state.scrollList.shift();
          state.distance = 0;
        }
      }, n * props.speed);
    };
    const go = (item) => {
      emit("click", item);
    };
    const handleClickIcon = () => {
      if (props.closeMode) {
        state.showNoticebar = !props.closeMode;
      }
      emit("close", state.scrollList[0]);
    };
    onMounted(() => {
      if (props.direction == "vertical") {
        if (slots.default) {
          updateSlotChild();
          watchSlots();
        } else {
          state.scrollList = [].concat(props.list);
        }
        setTimeout(() => {
          props.complexAm ? startRoll() : startRollEasy();
        }, props.standTime);
      } else {
        initScrollWrap(props.text);
      }
    });
    const updateSlotChild = () => {
      if (slots.default)
        state.scrollList = [].concat(slots.default()[0].children);
    };
    const watchSlots = () => {
      setTimeout(() => {
        var observer = new MutationObserver((slots2) => {
          state.showNotica = false;
          setTimeout(() => {
            state.showNotica = true;
          });
          updateSlotChild();
        });
        const ele = document.getElementsByClassName("nut-noticebar-custom-item")[0];
        if (ele) {
          observer.observe(ele, {
            childList: true,
            subtree: true
          });
        }
      }, 100);
    };
    onActivated(() => {
      if (state.keepAlive) {
        state.keepAlive = false;
      }
    });
    onDeactivated(() => {
      state.keepAlive = true;
      clearInterval(state.timer);
    });
    onUnmounted(() => {
      clearInterval(state.timer);
    });
    return {
      ...toRefs(props),
      ...toRefs(state),
      isEllipsis,
      classes,
      barStyle,
      contentStyle,
      horseLampStyle,
      wrap,
      content,
      handleClick,
      onClickIcon,
      onAnimationEnd,
      go,
      handleClickIcon,
      slots,
      pxCheck,
      wrapContentClass
    };
  }
});
const _hoisted_1 = { class: "nut-noticebar__page-lefticon" };
const _hoisted_2 = {
  ref: "wrap",
  class: "nut-noticebar__page-wrap"
};
const _hoisted_3 = { class: "showNotica" };
const _hoisted_4 = { class: "nut-noticebar-custom-item" };
const _hoisted_5 = ["onClick"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_Notice = resolveComponent("Notice");
  const _component_CircleClose = resolveComponent("CircleClose");
  const _component_ScrollItem = resolveComponent("ScrollItem");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes)
  }, [
    _ctx.direction == "across" ? withDirectives((openBlock(), createElementBlock("view", {
      key: 0,
      class: normalizeClass(["nut-noticebar__page", {
        "nut-noticebar__page--withicon": _ctx.closeMode,
        "nut-noticebar__page--close": _ctx.closeMode,
        "nut-noticebar__page--wrapable": _ctx.wrapable
      }]),
      style: normalizeStyle(_ctx.barStyle),
      onClick: _cache[3] || (_cache[3] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
    }, [
      createElementVNode("view", _hoisted_1, [
        renderSlot(_ctx.$slots, "left-icon", {}, () => [
          _ctx.leftIcon ? (openBlock(), createBlock(_component_Notice, {
            key: 0,
            size: "16px"
          })) : createCommentVNode("", true)
        ])
      ]),
      createElementVNode("view", _hoisted_2, [
        createElementVNode("view", {
          ref: "content",
          class: normalizeClass(_ctx.wrapContentClass),
          style: normalizeStyle(_ctx.contentStyle),
          onAnimationend: _cache[0] || (_cache[0] = (...args) => _ctx.onAnimationEnd && _ctx.onAnimationEnd(...args)),
          "on:webkitAnimationEnd": _cache[1] || (_cache[1] = (...args) => _ctx.onAnimationEnd && _ctx.onAnimationEnd(...args))
        }, [
          renderSlot(_ctx.$slots, "default", {}, () => [
            createTextVNode(toDisplayString(_ctx.text), 1)
          ])
        ], 38)
      ], 512),
      _ctx.closeMode || _ctx.$slots["right-icon"] ? (openBlock(), createElementBlock("view", {
        key: 0,
        class: "nut-noticebar__page-righticon",
        onClick: _cache[2] || (_cache[2] = withModifiers((...args) => _ctx.onClickIcon && _ctx.onClickIcon(...args), ["stop"]))
      }, [
        _ctx.$slots["right-icon"] ? renderSlot(_ctx.$slots, "right-icon", { key: 0 }) : (openBlock(), createBlock(_component_CircleClose, { key: 1 }))
      ])) : createCommentVNode("", true)
    ], 6)), [
      [vShow, _ctx.showNoticebar]
    ]) : createCommentVNode("", true),
    _ctx.scrollList.length > 0 && _ctx.direction == "vertical" && _ctx.showNoticebar ? (openBlock(), createElementBlock("view", {
      key: 1,
      class: "nut-noticebar__vertical",
      style: normalizeStyle(_ctx.barStyle)
    }, [
      _ctx.slots.default ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
        createElementVNode("view", {
          class: "nut-noticebar__vertical-list",
          style: normalizeStyle(_ctx.horseLampStyle)
        }, [
          createElementVNode("div", _hoisted_3, [
            (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.scrollList, (item, index2) => {
              return openBlock(), createBlock(_component_ScrollItem, {
                key: index2,
                style: normalizeStyle({ height: _ctx.height + "px", "line-height": _ctx.height + "px" }),
                item
              }, null, 8, ["style", "item"]);
            }), 128))
          ])
        ], 4),
        createElementVNode("view", _hoisted_4, [
          renderSlot(_ctx.$slots, "default")
        ])
      ], 64)) : (openBlock(), createElementBlock("ul", {
        key: 1,
        class: "nut-noticebar__vertical-list",
        style: normalizeStyle(_ctx.horseLampStyle)
      }, [
        (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.scrollList, (item, index2) => {
          return openBlock(), createElementBlock("li", {
            class: "nut-noticebar__vertical-item",
            key: index2,
            style: normalizeStyle({ height: _ctx.pxCheck(_ctx.height), lineHeight: _ctx.pxCheck(_ctx.height) }),
            onClick: ($event) => _ctx.go(item)
          }, toDisplayString(item), 13, _hoisted_5);
        }), 128))
      ], 4)),
      createElementVNode("view", {
        class: "go",
        onClick: _cache[4] || (_cache[4] = ($event) => !_ctx.slots.rightIcon && _ctx.handleClickIcon())
      }, [
        renderSlot(_ctx.$slots, "right-icon", {}, () => [
          _ctx.closeMode ? (openBlock(), createBlock(_component_CircleClose, {
            key: 0,
            color: _ctx.color,
            size: "11px"
          }, null, 8, ["color"])) : createCommentVNode("", true)
        ])
      ])
    ], 4)) : createCommentVNode("", true)
  ], 2);
}
const index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export {
  index as default
};
