// @nutui/nutui/dist/packages/_es/SideNavbarItem.js
import { computed, openBlock, createElementBlock, normalizeClass, withModifiers, createElementVNode, toDisplayString } from "vue";
import { c as createComponent } from "./component-81a4c1d0.js";
import { _ as _export_sfc } from "./_plugin-vue_export-helper-cc2b3d55.js";
import "../locale/lang";
const { componentName, create } = createComponent("side-navbar-item");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: ""
    },
    ikey: {
      type: String,
      default: ""
    }
  },
  emits: ["click"],
  setup: (props, context) => {
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    const handleClick = () => {
      context.emit("click");
    };
    return {
      classes,
      handleClick
    };
  }
});
const _hoisted_1 = ["ikey"];
const _hoisted_2 = { class: "nut-side-navbar-item__title" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    onClick: _cache[0] || (_cache[0] = withModifiers((...args) => _ctx.handleClick && _ctx.handleClick(...args), ["stop"])),
    ikey: _ctx.ikey
  }, [
    createElementVNode("span", _hoisted_2, toDisplayString(_ctx.title), 1)
  ], 10, _hoisted_1);
}
const index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export {
  index as default
};
