// @nutui/nutui/dist/packages/_es/SubSideNavbar.js
import { reactive, computed, onMounted, toRefs, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, withModifiers, toDisplayString, createBlock, withDirectives, normalizeStyle, renderSlot, vShow } from "vue";
import { c as createComponent } from "./component-81a4c1d0.js";
import { ArrowDown2, ArrowUp2 } from "@nutui/icons-vue";
import { _ as _export_sfc } from "./_plugin-vue_export-helper-cc2b3d55.js";
import "../locale/lang";
const { componentName, create } = createComponent("sub-side-navbar");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: ""
    },
    ikey: {
      type: [String, Number],
      default: ""
    },
    open: {
      type: Boolean,
      default: true
    }
  },
  components: { ArrowDown2, ArrowUp2 },
  emits: ["title-click"],
  setup: (props, context) => {
    const state = reactive({
      direction: ""
    });
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    const style = computed(() => {
      return {
        height: !state.direction ? "auto" : "0px"
      };
    });
    const handleClick = () => {
      context.emit("title-click");
      state.direction = !state.direction ? "up" : "";
    };
    onMounted(() => {
      state.direction = props.open ? "" : "up";
    });
    return {
      ...toRefs(state),
      classes,
      style,
      handleClick
    };
  }
});
const _hoisted_1 = ["ikey"];
const _hoisted_2 = { class: "nut-sub-side-navbar__title__text" };
const _hoisted_3 = { class: "nut-sub-side-navbar__title__icon" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_ArrowDown2 = resolveComponent("ArrowDown2");
  const _component_ArrowUp2 = resolveComponent("ArrowUp2");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    ikey: _ctx.ikey
  }, [
    createElementVNode("view", {
      class: "nut-sub-side-navbar__title",
      onClick: _cache[0] || (_cache[0] = withModifiers((...args) => _ctx.handleClick && _ctx.handleClick(...args), ["stop"]))
    }, [
      createElementVNode("span", _hoisted_2, toDisplayString(_ctx.title), 1),
      createElementVNode("span", _hoisted_3, [
        !_ctx.direction ? (openBlock(), createBlock(_component_ArrowDown2, { key: 0 })) : (openBlock(), createBlock(_component_ArrowUp2, { key: 1 }))
      ])
    ]),
    withDirectives(createElementVNode("view", {
      class: normalizeClass(["nut-sub-side-navbar__list", !_ctx.direction ? "nutFadeIn" : "nutFadeOut"]),
      style: normalizeStyle(_ctx.style)
    }, [
      renderSlot(_ctx.$slots, "default")
    ], 6), [
      [vShow, !_ctx.direction]
    ])
  ], 10, _hoisted_1);
}
const index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export {
  index as default
};
